File: LeetCode/LeetCode 39. 组合总和 (dfs).py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans = list()
        ss = list()

        def dfs(x: int, cnt: int, target: int, candidates: List[int]):
            if cnt == target:
                ans.append(ss[:])
                return
            
            for i in range(x, len(candidates)):
                if cnt + candidates[i] > target:
                    break;
                ss.append(candidates[i])
                dfs(i, cnt + candidates[i], target, candidates)
                ss.pop()
            
            return
        
        candidates.sort()
        dfs(0, 0, target, candidates)
        return ans

File: LeetCode/LeetCode 39. 组合总和 (dfs).cpp
class Solution {
public:
    vector <int> ss;
    vector <vector<int>> ans;

    void dfs(int n, int sum, int target, vector<int>& candidates) {
        if(sum == target) {
            ans.push_back(ss);
            return ;
        }
        for(int i = n; i < candidates.size(); i++) {
            if(sum + candidates[i] > target) break;
            ss.push_back(candidates[i]);
            dfs(i, sum + candidates[i], target, candidates);
            ss.pop_back();
        }
        return ;
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(0, 0, target, candidates);
        return ans;
    }
};